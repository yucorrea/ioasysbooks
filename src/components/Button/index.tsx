import React, { useEffect } from 'react';
import { TouchableOpacityProps, ActivityIndicator } from 'react-native';
import styled, { useTheme } from 'styled-components/native';

import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
} from 'react-native-reanimated';
interface Props extends TouchableOpacityProps {
  title?: string;
  loading: boolean;
}

export function Button({ title, loading = true, ...rest }: Props) {
  const { colors } = useTheme();

  const width = useSharedValue(86);
  const height = useSharedValue(36);
  const textOpacity = useSharedValue(1);

  useEffect(() => {
    if (loading) {
      width.value = withTiming(36, {duration: 200 });
      textOpacity.value = withTiming(0, { duration: 200 });
    }else {
      width.value = withTiming(86, { duration: 200 });
      textOpacity.value = withTiming(1, { duration: 200 });
    }
  }, [loading]);

  const buttonStyles = useAnimatedStyle(() => {
    return {
      width: width.value,
      height: height.value,
    };
  });

  const textStyles = useAnimatedStyle(() => {
    return { opacity: textOpacity.value };
  });

  return (
    <Animated.View style={buttonStyles}>
     <StyledContainer style={buttonStyles} disabled={loading} {...rest}>
        {loading ? (
          <ActivityIndicator  size={24} color={colors.primary}/>
        ) : (
          <StyledButtonText style={textStyles}>{title}</StyledButtonText>
        )}
      </StyledContainer>
    </Animated.View>
  );
}

const StyledContainer = styled.TouchableOpacity`
  justify-content: center;
  align-items: center;
  border-radius: 44px;
  background: ${({ theme }) => theme.colors.shape};
`;

const StyledButtonText = styled(Animated.Text)`
  color: ${({ theme }) => theme.colors.primary};
  font-size: 16px;
  font-family: ${({ theme }) => theme.fonts.medium};
`;
