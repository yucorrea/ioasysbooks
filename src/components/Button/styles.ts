import styled from 'styled-components/native';

export const Container = styled.TouchableOpacity`
  width: 85px;
  height: 36px;
  background: ${({ theme }) => theme.colors.background};

  border-radius: 44px;
  justify-content: center;
  align-items: center;
`;

export const ButtonText = styled.Text`
  color: ${({ theme }) => theme.colors.primary};
  font-size: 16px;
  font-family: ${({ theme }) => theme.fonts.medium};
`;
