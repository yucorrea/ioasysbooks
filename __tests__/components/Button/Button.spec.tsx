import React from 'react';
import { fireEvent } from '@testing-library/react-native';

import Button from './../../../src/components/Button';
import {
  matchSnapshotWithTheme,
  renderWithTheme,
} from '../../../src/helpers/functions/testing';

describe('Buttom Component', () => {

  it('matches snapshots', () => {
    const snap = matchSnapshotWithTheme(<Button title="Click-me" />);
    expect(snap).toMatchSnapshot();
  });

  it('render correctly', () => {
    const mockFunction = jest.fn();
    const { getByTestId } = renderWithTheme(
      <Button onPress={mockFunction} testID="button-id" title="Click-me" />,
    );
    const button = getByTestId('button-id');
    fireEvent.press(button);
    expect(mockFunction).toHaveBeenCalled();
  });

  it('render button loading if props has true', () => {
     renderWithTheme( <Button onPress={jest.fn} title="Click-me" loading /> );
  });
});
