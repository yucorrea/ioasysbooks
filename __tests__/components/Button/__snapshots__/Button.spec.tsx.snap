// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttom Component matches snapshots 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "height": 36,
        "width": 86,
      },
    }
  }
  collapsable={false}
  style={
    Object {
      "height": 36,
      "width": 86,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "animatedStyle": Object {
          "current": Object {},
        },
        "backgroundColor": "#ffffff",
        "borderRadius": 44,
        "initial": Object {
          "updater": [Function],
          "value": Object {
            "height": 36,
            "width": 86,
          },
        },
        "justifyContent": "center",
        "opacity": 1,
        "viewDescriptors": Object {
          "add": [Function],
          "batchToRemove": Object {},
          "items": Array [
            Object {
              "name": undefined,
              "tag": undefined,
            },
          ],
          "rebuildsharableViewDescriptors": [Function],
          "remove": [Function],
          "sharableViewDescriptors": Object {
            "_animation": null,
            "_id": 6,
            "_listeners": Array [],
            "_setter": [Function],
            "_value": Array [],
          },
          "tags": Object {},
          "waitForInsertSync": false,
          "waitForRemoveSync": false,
        },
        "viewsRef": Object {
          "add": [Function],
          "items": Object {},
          "remove": [Function],
        },
      }
    }
  >
    <Text
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 1,
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "color": "#B22E6F",
          "fontFamily": "Heebo-Medium",
          "fontSize": 16,
          "opacity": 1,
        }
      }
    >
      Click-me
    </Text>
  </View>
</View>
`;

exports[`Buttom Component matches snapshots 2`] = `undefined`;
